File: test/gif-pixels/giffer.c
# include <stdio.h>
# include <stdlib.h>
# include <time.h>

# define WIDTH 50
# define HEIGHT 50
# define PIXELS WIDTH * HEIGHT

struct color {
  unsigned char r;
  unsigned char g;
  unsigned char b;
  unsigned char a;
};

struct point {
  unsigned int x;
  unsigned int y;
};

static const unsigned int color_distance = 20;
static struct point stack [(WIDTH+2)*(HEIGHT+2)];
static unsigned int sp = 0;
static const struct point max = {WIDTH,HEIGHT};
static struct color raster [WIDTH][HEIGHT];

struct color near ( struct color c) {
  c.r += (rand() % color_distance)-(color_distance/2);
  c.g += (rand() % color_distance)-(color_distance/2);
  c.b += (rand() % color_distance)-(color_distance/2);
  c.a = 0;
  return c;
}

void printint(unsigned int a)
{
  putchar((unsigned char)a);
  putchar((unsigned char)(a>>8));
}


void write(unsigned int x,unsigned int y,struct color c) {
  raster[x][y] = c;
  /*fprintf(stderr,
    "%d %d #%02x%02x%02x\n",
    x,y,
    c.r,
    c.g,
    c.b);*/
  /* Graphic Control Extension Block */
  putchar(0x21);
  putchar(0xf9);
  putchar(0x04);
  putchar(0x00);
  putchar(0x01);
  putchar(0x00);
  putchar(0x00);
  putchar(0x00);
  /* Image Block */
  putchar(0x2c);
  printint(x);
  printint(y);
  putchar(0x01);
  putchar(0x00);
  putchar(0x01);
  putchar(0x00);
  putchar(0x80);
  putchar(c.r);
  putchar(c.g);
  putchar(c.b);
  putchar(0x00);
  putchar(0x00);
  putchar(0x00);
  putchar(0x02);
  putchar(0x02);
  putchar(0x44);
  putchar(0x01);
  putchar(0x00);

}
void walk(struct point p, struct point max) {
  if (p.x > 0) {
    if (raster[p.x-1][p.y].a) {
      write(p.x-1, p.y, near(raster[p.x][p.y]));
      stack[sp++] = (struct point){p.x-1,p.y};
    }
  }
  if (p.y > 0) {
    if (raster[p.x][p.y-1].a) {
      write(p.x, p.y-1, near(raster[p.x][p.y]));
      stack[sp++] = (struct point){p.x,p.y-1};
    }
  }
  if (p.x < max.x) {
    if (raster[p.x+1][p.y].a) {
      write(p.x+1, p.y, near(raster[p.x][p.y]));
      stack[sp++] = (struct point){p.x+1,p.y};
    }
  }
  if (p.y < max.y) {
    if (raster[p.x][p.y+1].a) {
      write(p.x, p.y+1, near(raster[p.x][p.y]));
      stack[sp++] = (struct point){p.x,p.y+1};
    }
  }
}
void gif_header(void) {
  /* header */
  printf("GIF89a");
  /* logical screen info */
  printint(WIDTH);
  printint(HEIGHT);
  putchar(0x70); /* R8G8B8 no palette */
  putchar(0x00); /* background color index (meaningless) should be 0x00*/
  putchar(0x00); /* Pixel Aspect Ratio (no information) */

  putchar(0x21); /* Extension */
  putchar(0xff);
  putchar(0x0b);
  printf("NETSCAPE2.0");
  putchar(0x03);
  putchar(0x01);
  putchar(0x01);
  putchar(0x00);
  putchar(0x00);
}

int main(int argc, char * argv[]) {
  unsigned int i;
  unsigned int j;
  struct point x;
  unsigned int cont = 1;
  for (i=0;i<WIDTH;i++) {
    for (j=0;j<HEIGHT;j++) {
      raster[i][j] = (struct color){0x80,0x80,0x80,0xff};
    }
  }
  srand(time(NULL));
  stack[sp++] = (struct point){rand()%WIDTH,rand()%HEIGHT};

  gif_header();

  while (cont) {
    i = 0;
    j = (unsigned int)rand() % sp;
    x = stack[j];
    while (x.x == WIDTH && x.y == HEIGHT) {
      j = (j+1)%sp;
      x = stack[j];
      i++;
      if (i>PIXELS) {
        putchar(';'); /* gif trailer */
        return EXIT_SUCCESS;
      }
    }
    stack[j] = max;
    walk(x,max);
  }
  putchar(';'); /* gif trailer */
  return EXIT_SUCCESS;
}

File: test/gif-pixels/giffer.py
#!/usr/bin/env python2.7
# coding: utf-8

import sys
import struct
import random

class Gif():
  def __init__(self,width=500,height=500,color=(0,0,0)):
    self.width = width
    self.height = height
    self.canvas_color = color
    self.frames = []
    
  def write(self):
    #with open(self.filename,'w') as ofi:
    #  ofi.write(self.get_header())
    #  ofi.write('lol')
    #  ofi.close()
    sys.stdout.write(self.header())
    for f in self.frames:
      sys.stdout.write(f.get())
    sys.stdout.write(';')


  def header(self):
    scr = struct.pack('<6sHHBBB',
      'GIF89a',
      self.width,
      self.height,
      0b01110000, # no global color palette, 0b111 = 7 +1= 8bit color depth, 0b0000 = no palette -> no information on palette
      0,          # background color index
      0)          # pixel aspect ratio (0 = no information)

    # http://odur.let.rug.nl/~kleiweg/gif/netscape.html
    ext = struct.pack('<3s11s2sHB','\x21\xff\x0b','NETSCAPE2.0','\x03\x01',1,0)

    return scr+ext

class Frame():
  def __init__(self,x,y,color,w=1,h=1):
    self.x = x
    self.y = y
    self.h = h
    self.w = w
    self.color = color

  def get(self):
    #header = struct.pack('<BHHHHB',0x2c,self.position.x,self.position.y,1,1,0x87)
    #graphics = struct.pack('<3sBHBB','\x21\xf9\x04',0,duration,0,0)
    return ''.join([
      '\x21\xf9\x04\x00\x01\x00\x00\x00\x2c',
      struct.pack('<HH',self.x,self.y),
      struct.pack('<HH',self.w,self.h),
      '\x80',
      ''.join([chr(i) for i in self.color]),
      '\x00\x00\x00\x02\x02\x44\x01\x00'])


def randomcolor():
  return (random.randrange(256),random.randrange(256),random.randrange(256))

if __name__ == '__main__':
  width   = 40
  height  = 40
  a = Gif(width,height)
  a.frames.append(Frame(width,height,(0,0,0),True))
  for i in xrange(width*height):
    a.frames.append(Frame(random.randrange(width),
                          random.randrange(height),
                          randomcolor(),
                          ))
  a.write()
